dfinity/motoko
# Streaming `Stabilization` (emission)

## Purpose

Streaming stabilisation is a necessary prerequisite for the compacting GC. This is explained in #2909 and #2994.

## Approach

We build (and extend) on the `Stream` abstraction introduced in #3161 and instantiate a new module that is capable of caching streaming data into a blob. When the cache limit is reached, an outputter bottleneck function will be called to move out byte ranges to a backing store. This is `BlobStream`.

`Stabilization` then further extends this idiom to set up the bottleneck function to directly write big chunks into stable memory.

Fixes #2909.

## Fine points

This section details some of the more tricky issues that arose while implementation, and how they have been mitigated. With some more thought, these may be completely eliminated in the future.

### Linking against `ic::stable64_write` from Rust

The stream bottleneck is also defined in Rust (in `stream.rs`), so it wants to call `stable64_write` from the `ic` module. This  will cause linker errors for the WASI mode. Instead we call `stable64_write_moc` from `env` (which `moc` provides). In turn `moc` will generate a trapping `stable64_write_moc` for WASI mode or re-export the proper `stable64_write`.

### The contract of `Stream::reserve`

When calling `Stream::reserve` it is the caller's responsibility to immediately fill in the reserved region. Note that the returned pointer is unaligned, so either write bytes or use `align = 0` in Wasm.

### Finalising a `Stream` (OCaml)

Some streams like to write out bytes eagerly, while others may need to cache all bytes until `Stream.terminate`. At this point the length of _all_ written bytes is returned, along with an optional pointer to the start of the data. For eager streams it might be impossible to return a pointer to the 0th byte, because it is not in the cache any more. In this case `0` must be returned. Should a value `!= 0` be returned, then the stream's user has the obligation to consume the bytes (freeing is _not_ necessary/possible). A helper is provided from the `Stream` API: `Stream.finalize_buffer` takes a code snippet that the user provides that consumes the length/dataptr, but the `Stream` implementation will decide whether the code will be inserted or ignored (because all bytes are already at their destination, e.g. in the backing store).
To be fully `Stream`-agnostic, either the dataptr returned from `Stream.terminate` must be checked at runtime, or the compile-time `Stream.finalize_buffer` facility must be used.

-------
## TODOs

Not all of these are strictly necessary (at this point), but nice to have.

- [x] Move boring refactors in a prerequisite PR! — #3182
- more `Stream` tests
- [x] outputting to stable
  - track totals (see `absolute_offset`)
  - support new SM layout too
- check every non-constant `stream_reserve` for `blob` overflow
  - is overflow bad here?
  - currently benign, but I left `FIXME`s in place – follow-on PR will be written
- [ ] consider moving a BigNum -> (S)LEB128 converter into `Stream` to alleviate the "`reserve` by unknown amount" issue
- make `as_blob` return a `*const` (others too?)
- [ ] [define and use](https://github.com/dfinity/motoko/pull/3182#discussion_r843605469) `compile_extend64`
- Should we use different `Stream`s for `Externalization` depending on the GC strategy?
- [ ] decouple the `StableMemoryStream` from the `N+4` start address
